### PR TITLE
YOLOE Visual+Text Prompt

### LINKED ISSUE #21943: Visual + Text Prompting
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

So i have been playing with YOLOE text and visual prompting , issue that i generally faced are that text-prompting doesn't always give model the capability of visual features , and visual prompting gives only the visual features  which doesn't necessarily always represent the class  , so  i was wondering since we are converting text prompts into embedding and visual prompt into embeddings as well , can't we just pass them along with each other ,and if not  ,then how about we pass them by doing a cross-attention mechanism between the embeddings of text and visual prompt embedding. I mean this would be revolutionary if it works out , Are you guys already working on such feature or do you find this interesting?

### Use case

Giving visual and text prompt will give model ability to interpret the class features really well.

### Additional

I have been exploring online for this but didn't really find anything useful , maybe this can be a somewhat okayish reference : https://github.com/intel/TVP?tab=readme-ov-file

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by harshm2601**
This pull request adds multimodal fusion support to YOLOE, enabling the model to combine text and visual prompt embeddings for improved object detection. It introduces new fusion modes, updates inference logic to handle these modes, provides comprehensive documentation, and includes thorough unit tests to ensure correct behavior and error handling.

**Multimodal Fusion Feature:**

* Added `fuse_tpe_vpe` function in `ultralytics/models/yolo/yoloe/fusion.py` to support concatenation and weighted sum fusion of text and visual prompt embeddings, including handling for mismatched counts and custom mapping.
* Updated `YOLOEVPSegPredictor.inference` in `ultralytics/models/yolo/yoloe/predict.py` to accept new fusion parameters (`fuse_prompts`, `fuse_alpha`, `fuse_map`) and perform the appropriate fusion logic during inference. [[1]](diffhunk://#diff-0a6dd19491d3eb71d20ce7875e24aa05b3ae0b19ae7a0139de103e55f0198a13L135-R135) [[2]](diffhunk://#diff-0a6dd19491d3eb71d20ce7875e24aa05b3ae0b19ae7a0139de103e55f0198a13R145-R162)

**Documentation:**

* Added `docs/en/howto/yoloe_multimodal.md` with a detailed guide on using multimodal fusion, including API usage, parameter explanations, examples, and implementation notes.

**Testing:**

* Introduced `tests/test_yoloe_multimodal.py` with unit tests covering all fusion modes, edge cases, and error handling for the new fusion logic.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Adds multimodal fusion to YOLOE, letting you combine text and visual prompts for more robust detection, with clear docs, tests, and a lightweight API. 🤝🧠

### 📊 Key Changes
- New multimodal how-to guide with examples and parameters: “YOLOE: Combine Text and Visual Prompts” 📘
- Fusion utility added: `fuse_tpe_vpe` supports:
  - Concatenation mode (“concat”) to run text and visual prompts together with zero extra cost
  - Weighted-sum mode (“sum”) with `fuse_alpha` and optional `fuse_map`, including normalization for stability ⚖️
- YOLOE predictor updated to accept:
  - `fuse_prompts` ("concat" | "sum"), `fuse_alpha`, and `fuse_map`
  - Two-pass inference for “sum” mode (extract VPE, then fuse with TPE)
  - Backward-compatible default (visual prompts only) 🧩
- Comprehensive tests covering fusion behaviors, edge cases, and error handling ✅

### 🎯 Purpose & Impact
- Better results through multimodal prompting: combines semantic context (text) with visual specificity (visual prompts) for stronger detections 🎯
- Flexible performance trade-offs:
  - “Concat”: zero overhead, independent class banks
  - “Sum”: small overhead for improved robustness
- Safer and clearer usage with explicit errors for mismatched sizes, invalid modes, or missing maps ⚠️
- Smooth adoption: existing YOLOE workflows remain unchanged unless fusion is enabled, with simple, well-documented APIs 🚀

Closes #21943

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @harshm2601, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline the review. An Ultralytics engineer will also assist you here shortly. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

Additional tips for this PR:
- Please confirm backward compatibility for existing YOLOE flows when fusion is not enabled
- Ensure new arguments (`fuse_prompts`, `fuse_alpha`, `fuse_map`) are documented consistently and have sensible defaults
- Verify naming consistency and error messages are clear and actionable
- If available, add brief notes in the PR description on expected performance or latency trade-offs for “concat” vs “sum” modes

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀🧠

**Review by NeelDevenShah — COMMENTED** [review]
Looks like a great idea to combine both embeddings into a fused one — really like the approach. I went through the README and code, and everything seems clear, consistent, and ready to go.

**Review by harshm2601 — COMMENTED** [review]
.

**Comment by rogervendrell** [comment]
Hello everyone, I am also interested in fusing both types of embeddings into a single representation.
My concern is that, as mentioned in Issue #21479, simply averaging prototypes dilutes the signal of each embedding and ends up providing worse performance.
I assume that if it happens with multi-image prompts it is likely to happen as well when mixing textual with visual prompts. This is why training attention weights seems reasonable.

My questions are:

1. What would be the natural insertion point for this modification?
2. Training would be limited to image+text pair datasets. Also a computational overhead would be introduced, but the model could be frozen to some extent as proposed with Linear Probing.

I'd like to help by submitting a PR if needed.
Context: I am working on wrapping YOLOE with a Continual Learning framework as a proof-of-concept for testing how visual+textual explanations could help the model perform better in the long-term.

File: docs/en/howto/yoloe_multimodal.md [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+# 1) Load model and set text classes once
+model = YOLOE("yoloe-11l-seg.pt")
+names = ["person", "bus"]
+model.set_classes(names, model.get_text_pe(names))
+
+# 2) Build visual prompts (same as existing API)
+visual_prompts = dict(
+    bboxes=[np.array([[221.5, 405.8, 344.9, 857.5]])],
+    cls=[np.array([0])],  # map prompt to class 0 ("person")
+)
```
Comment (Vinyzu) [inline docs/en/howto/yoloe_multimodal.md:41]: This should be:
```suggestion
visual_prompts = dict(
    bboxes=np.array([[221.5, 405.8, 344.9, 857.5]]),
    cls=np.array([0]),  # map prompt to class 0 ("person")
)
```

File: docs/en/howto/yoloe_multimodal.md [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+# 3) Run with both prompts - Concatenation Mode
+#    Treats text and visual as separate class banks (zero overhead)
+results = model.predict(
+    "ultralytics/assets/zidane.jpg",
+    visual_prompts=visual_prompts,
```
Comment (Vinyzu) [inline docs/en/howto/yoloe_multimodal.md:47]: Either you missed adding a `refer_image` here, or youve set the wrong source image. The Visual Prompts are for `bus.jpg`.
```suggestion
    visual_prompts=visual_prompts,
    refer_image="ultralytics/assets/bus.jpg",
```

**Review by Vinyzu — COMMENTED** [review]
The main problem with this approach is that the YOLOEVPSegPredictor will get reset [after vpe evaluation](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/model.py#L443).
Then, YoloE will [load](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/engine/model.py#L549) the normal SegmentationPredictor as per [task_map](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/model.py#L261), which will then just use the VPE. So displayed results are the same without fusing prompts, as you will just receive the same results as a run with just visual prompting.

**Comment by Vinyzu** [comment]
There are also more fundamental issues with the approach, e.g. trying to access the `pe` to get the [text prompt encodings/tpe](https://github.com/harshm2601/ultralytics/blob/main/ultralytics/models/yolo/yoloe/predict.py#L150), but these tpe being overwritten by the vpe in the [refer_image inference call](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/models/yolo/model.py#L440-L441).
So we would probably have to store the `tpe`, restore it and then fuse it with the vpe

@harshm2601 If i can help with this PR let me know. I have fixed some of these issues locally, but still not there with a working solution yet. (For example because of this `tpe` overwriting issue)

**Comment by harshm2601** [comment]
Hey @Vinyzu , can you checkout changes i made in fusion.py and predict.py , which will help you get better idea of what's happening and if you find any issue , please comment along with code as well  like you did earlier , Thanks!

**Comment by Vinyzu** [comment]
Ive checked/reviewed these changes and i think they look good. The problem i see isnt in these files, but mainly in `models/yolo/model.py`.
See my comments above for some of the issues i found.

**Comment by Vinyzu** [comment]
@harshm2601 In what cases would Concatenation Mode make sense?
I feel like Concat Mode, unlike Sum Mode, makes no effort in actually solving any of the issues its designed to solve, but may actually decrease performance by blurring the semantic understanding the text prompts provide and the visual nuances the visual prompts provide.

**Comment by harshm2601** [comment]
@Vinyzu Hey , you are right  , it doesn't make sense  , but i just added that as an option as i thought it might make sense in some use cases that are not known now
.

**Comment by glenn-jocher** [comment]
Good point, and I’m aligned with @Vinyzu here: since YOLOE already combines text and visual prompts internally via `YOLOEModel.get_cls_pe`/`get_vpe`, adding an explicit “concat” fusion mode without a concrete use case or ablation is likely to create confusion and extra maintenance, so for any follow‑up it would be better to focus on a single, well‑justified fusion strategy (e.g. weighted sum) and on fixing the `tpe`/`vpe` lifecycle around `refer_image` in `models/yolo/model.py`; this PR is closed now, but if you iterate on that design with benchmarks or qualitative examples, please open a new PR referencing this thread so the team can review, and you can use the [YOLOE documentation](https://docs.ultralytics.com/models/yoloe/) as the canonical behavior reference.
